src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,456 @@
+        uint256 fees = 0;
+        if (assetBalance > daiDebt) {
+            fees = assetBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
+            }
+        }
+
+        if (amount == 0 && fees == 0) {
+            emit Unwind(ilk, 0);
+            return;
+        }
+
+        require(amount <= 2 ** 255, "DssDirectDepositHub/overflow");
+
+        // To save gas you can bring the fees back with the unwind
+        uint256 total = _add(amount, fees);
+        pool.withdraw(total);
+        daiJoin.join(address(this), total);
+
+        // normalized debt == erc20 DAI to pool (Vat rate for this ilk fixed to 1 RAY)
+
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(pool), address(pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(fees, RAY));
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        } else {
+            // This can be done with the assumption that the price of 1 aDai equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, address(end_), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        }
+
+        emit Unwind(ilk, amount);
+    }
+
+    function exec(bytes32 ilk_) external {
+        D3MPoolLike pool = ilks[ilk_].pool;
+
+        pool.accrueIfNeeded();
+        uint256 availableAssets = pool.maxWithdraw();
+        uint256 currentAssets = pool.assetBalance();
+
+        if (vat.live() == 0) {
+            // MCD caged
+            require(end.debt() == 0, "DssDirectDepositHub/end-debt-already-set");
+            require(ilks[ilk_].culled == 0, "DssDirectDepositHub/module-has-to-be-unculled-first");
+            _unwind(
+                ilk_,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                Mode.MCD_CAGED,
+                currentAssets
+            );
+        } else if (live == 0) {
+            // This module caged
+            _unwind(
+                ilk_,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                ilks[ilk_].culled == 1
+                ? Mode.MODULE_CULLED
+                : Mode.NORMAL,
+                currentAssets
+            );
+        } else {
+            // Normal path
+            uint256 targetAssets = ilks[ilk_].plan.getTargetAssets(currentAssets);
+
+            if (targetAssets > currentAssets) {
```

Comment:
Just realized there is a mistake here which is also present on the live D3M. Calling wind or unwind should not only depend on the plan's target, but also the debt ceiling. If we lower the debt ceiling below the current debt then this module should trigger an unwind instead of trying to wind (but failing due to the comment above).

Reply:
how do you feel about this solution: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/e48306da572402bfbcd30258e33b2bc787d35d48

Reply:
Yes, I like this.

Reply:
If we are going this path, shouldn't we also check the general debt ceiling?

Reply:
Btw I think the `line`s check should actually happen in an upper level (or actually in both parts of the conditions).
Let's say the comparison of `targetAssets` and `currentAssets` determines it needs to `unwind` but due the debt ceiling it should actually `unwind` even more.
So to make it very correct, we should get both `line`s, determine if we overpass the limit of any of them, and get that `gap` and use the max between that value and `currentAssets - targetAssets`.